### PR TITLE
feat/remove vulkan feature

### PR DESCRIPTION
**Pull request by zcorniere**

### CODE REVIEW

File: libgraphics/include/pivot/graphics/vk_utils.hxx [hunk 1]
```diff
@@ -29,7 +31,31 @@ constexpr void vk_try(vk::Result err)
...
-std::vector<std::byte> readFile(const std::string &filename);
+template <typename T = std::byte>
+std::vector<T> readBinaryFile(const std::filesystem::path &filename)
+{
+    size_t fileSize = std::filesystem::file_size(filename);
+    std::ifstream file(filename, std::ios::binary);
+    std::vector<T> fileContent(fileSize / sizeof(T));
+
+    if (!file.is_open()) throw std::runtime_error("failed to open file " + filename.string());
+    file.read(reinterpret_cast<char *>(fileContent.data()), fileSize * sizeof(T));
```
Comment (Ersikan) [inline libgraphics/include/pivot/graphics/vk_utils.hxx:42]: Check how many characters were read with `gcount`, and check that no error occured with `fail` and `bad`.

File: libgraphics/include/pivot/graphics/vk_utils.hxx [hunk 2]
```diff
@@ -29,7 +31,31 @@ constexpr void vk_try(vk::Result err)
...
+    file.close();
+    return fileContent;
+}
+
+/// Read a whole file into a string
+std::string readFile(const std::filesystem::path &filename);
+
+template <typename T = std::byte>
+std::size_t writeBinaryFile(const std::filesystem::path &filename, const std::span<const T> &code)
```
Comment (Ersikan) [inline libgraphics/include/pivot/graphics/vk_utils.hxx:51]: No need for the first const on span
```suggestion
std::size_t writeBinaryFile(const std::filesystem::path &filename, std::span<const T> &code)
```

File: libgraphics/source/vk_debug.cxx [hunk 3]
```diff
@@ -5,19 +5,26 @@ namespace pivot::graphics::vk_debug
...
 #ifndef NDEBUG
     vk::DebugUtilsLabelEXT markerInfo{
         .pLabelName = pMarkerName,
         .color = color,
     };
     cmdbuffer.beginDebugUtilsLabelEXT(markerInfo);
+#else
+    (void)cmdbuffer;
+    (void)pMarkerName;
+    (void)color;
```
Comment (Ersikan) [inline libgraphics/source/vk_debug.cxx:18]: [maybe_unused]

File: libgraphics/source/vk_utils.cxx [hunk 4]
```diff
@@ -7,18 +7,16 @@
...
-std::vector<std::byte> readFile(const std::string &filename)
+std::string readFile(const std::filesystem::path &filename)
 {
-    size_t fileSize = 0;
-    std::vector<std::byte> fileContent;
-    std::ifstream file(filename, std::ios::ate | std::ios::binary);
+    /// Must be opened in binary mode, so Windows won't mess with the newlines
+    std::string fileContent;
+    std::ifstream file(filename, std::ios::binary);
+    size_t fileSize = std::filesystem::file_size(filename);
```
Comment (Ersikan) [inline libgraphics/source/vk_utils.cxx:15]: TOCTOU

File: libgraphics/source/vk_utils.cxx [hunk 5]
```diff
@@ -7,18 +7,16 @@
...
 
-    if (!file.is_open()) { throw std::runtime_error("failed to open file " + filename); }
-    fileSize = file.tellg();
+    if (!file.is_open()) throw std::runtime_error("failed to open file " + filename.string());
     fileContent.resize(fileSize);
-    file.seekg(0);
-    file.read((char *)fileContent.data(), fileSize);
+    file.read(fileContent.data(), fileSize);
```
Comment (Ersikan) [inline libgraphics/source/vk_utils.cxx:19]: gcount, fail, bad. Like above

File: libpivot/include/pivot/engine.hxx [hunk 6]
```diff
@@ -43,7 +43,7 @@ protected:
     builtins::Camera m_camera;
     bool m_paused = true;
 
-    virtual void onTick(float delta){};
```
Comment (Ersikan) [inline libpivot/include/pivot/engine.hxx:46]: Keep the name ! It's useful to know what this float means

File: libscript/CMakeLists.txt [hunk 7]
```diff
@@ -9,6 +9,17 @@ add_library(
...
+    ${PROJECT_NAME}
+    PRIVATE $<$<OR:$<CXX_COMPILER_ID:Clang>,$<CXX_COMPILER_ID:AppleClang>,$<CXX_COMPILER_ID:GNU>>:
+            -Wall -Wextra>
+            $<$<CXX_COMPILER_ID:MSVC>:
+            /W4
+            /wd4267
+            /wd4201
+            /wd4244>
+)
+
```
Comment (Ersikan) [inline libscript/CMakeLists.txt:22]: Maybe put it in a function instead of duplicating it everywhere ?

**Review by Ersikan — CHANGES_REQUESTED** [review]

File: libgraphics/CMakeLists.txt [hunk 8]
```diff
@@ -34,14 +36,15 @@ if(NOT ktx_POPULATED)
...
-    add_subdirectory(${ktx_SOURCE_DIR} ${ktx_SOURCE_DIR})
+    add_subdirectory(${ktx_SOURCE_DIR} ${ktx_BINARY_DIR})
 endif()
 
 fetchcontent_declare(
     tinygltf
     GIT_REPOSITORY https://github.com/syoyo/tinygltf.git
     GIT_TAG 912856823dcb2badf2074055b6d006f7bb3800b8
 )
+fetchcontent_getproperties(tinygltf)
```
[the remark this answers is not in the document]
Comment (zcorniere) [reply libgraphics/CMakeLists.txt:47]: I forgot to add this when I set up the CMake at the beginning.

File: libscript/CMakeLists.txt (hunk 7 above)
Comment (zcorniere) [reply libscript/CMakeLists.txt:22]: fixed in #189

File: libgraphics/include/pivot/graphics/vk_utils.hxx [hunk 9]
```diff
@@ -29,7 +31,36 @@ constexpr void vk_try(vk::Result err)
...
+    assert(!std::filesystem::is_symlink(filename));
+    size_t fileSize = std::filesystem::file_size(filename);
+    std::ifstream file(filename, std::ios::binary);
+    std::vector<T> fileContent(fileSize / sizeof(T));
+
+    if (!file.is_open()) throw std::runtime_error("failed to open file " + filename.string());
+    file.read(reinterpret_cast<char *>(fileContent.data()), fileSize * sizeof(T));
+    assert(!file.bad());
+    assert(!file.fail());
+    assert(fileContent.size() == (unsigned long)file.gcount());
```
Comment (Ersikan) [inline libgraphics/include/pivot/graphics/vk_utils.hxx:47]: wtf

**Review by Ersikan — CHANGES_REQUESTED** [review]

**Review by Ersikan — APPROVED** [review]

**Pull request merged**